### PR TITLE
8256487: Handle disableEagerInitialization for archived lambda proxy classes

### PR DESCRIPTION
**Pull request by calvinccheung**
Before this change, the setting of the `jdk.internal.lambda.disableEagerInitialization` property was not captured during dumping of lambda proxy classes. There's a workaround in `LambdaProxyClassArchive.find`, it won't call `findFromArchive` if the above property is set.

This change adds handling of the `jdk.internal.lambda.disableEagerInitialization` property, specifically:

- remove the above workaround;

- capture the setting of the property in the archive header during CDS dump time;

- during runtime when finding an archived lambda proxy class, the setting of the property will be compared with the stored value in the archive header.
If the values don't match, the archived lambda proxy class won't be used.

Tests:

- [x] ran all cds tests locally on linux-x64

- [x] ran the `hotspot_appcds_dynamic` test group with `-Dtest.dynamic.cds.archive=true` on linux-x64

- [x] mach5 tiers 1,2,3 (in progress)

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed

### Issue
 * [JDK-8256487](https://bugs.openjdk.java.net/browse/JDK-8256487): Handle disableEagerInitialization for archived lambda proxy classes

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1301/head:pull/1301`
`$ git checkout pull/1301`

### CODE REVIEW

**Comment by cl4es** [comment]
I'm not sure if this is a good idea, TBH. The disableEagerInitialization setting is for native-image pre-generation purposes and the less CDS cares about it, the better. I'd prefer it if there's no trace of the property in hotspot sources.

**Comment by calvinccheung** [comment]
Hi Claes,

Thanks for taking a look.

So should I keep the following `!initialize` check in LambdaProxyClassArchive?
```
109         if (!loadedByBuiltinLoader(caller) || !initialize ||
110             !CDS.isSharingEnabled() || isSerializable || markerInterfaces.length > 0 || additionalBridges.length > 0)
111             return null;
```
If we keep the above code, I think we don't need to pass the `initialize` to `findFromArchive` and eventually to `JVM_LookupLambdaProxyClassFromArchive`.

Let me know if the above is what you have in mind?

thanks,
Calvin

**Comment by cl4es** [comment]
Right, I'd drop that argument - I would go further and suggest making calls to both `LambdaProxyClassArchive.register` and `LambdaProxyClassArchive.find` conditional on `disableEagerInitialization` being `false` to avoid any accidental mix-up and reduce complexity of these orthogonal features/concerns.

File: src/hotspot/share/memory/metaspaceShared.cpp [hunk 1]
```diff
@@ -73,6 +73,8 @@
 #include "gc/g1/g1CollectedHeap.inline.hpp"
 #endif
 
+#include <string.h>
```
Comment (yminqi) [inline src/hotspot/share/memory/metaspaceShared.cpp:76]: This include is strange, usually we do not include std head file in .cpp file.

File: src/hotspot/share/memory/metaspaceShared.cpp [hunk 2]
```diff
@@ -1815,6 +1818,14 @@ bool MetaspaceShared::use_full_module_graph() {
   return result;
 }
 
+void MetaspaceShared::set_disable_eager_init(const char* value) {
```
Comment (yminqi) [inline src/hotspot/share/memory/metaspaceShared.cpp:1821]: strcasecmp is not platform dependent, why not use it for all? It does not need <string.h> included.

File: test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/LambdaEagerInit.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+ * @test
+ * @bug 8256487
+ * @summary Run the LambdaEagerInitTest.java test in dynamic CDS archive mode.
+ *          Tests various combinations of base and dynamic archives dumping and
+ *          run with those archives with the -Djdk.internal.lambda.disableEagerInitialization
+ *          property set to {true, false}.
+ * @requires vm.cds
+ * @library /test/lib /test/hotspot/jtreg/runtime/cds/appcds
+ * @compile ../../../../../../jdk/java/lang/invoke/lambda/LambdaEagerInitTest.java
+ * @build sun.hotspot.WhiteBox
```
Comment (yminqi) [inline test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/LambdaEagerInit.java:35]: Do we need WB here?

**Comment by mlchung** [comment]
I agree with Claes that this is a wrong move to archive lambda proxies even if `disableEagerInitialization` is set.   A simple fix would be in:

```
private Class<?> spinInnerClass() throws LambdaConversionException {
      if (!disableEagerInitialization) {
          if (LambdaProxyClassArchive.isDumpArchive()) {
              ...
          }
          // load from CDS archive if present
         Class<?> archiveClass = LambdaProxyClassArchive.find(targetClass,
                                                           samMethodName,
                                                           invokedType,
                                                           samMethodType,
                                                           implMethod,
                                                           instantiatedMethodType,
                                                           isSerializable,
                                                           markerInterfaces,
                                                           additionalBridges,
                                                           !disableEagerInitialization);
           if (archiveClass != null)  return archiveClass;
      }
      return generateInnerClass();
}
```

File: src/hotspot/share/memory/metaspaceShared.cpp (hunk 1 above)
Comment (calvinccheung) [reply src/hotspot/share/memory/metaspaceShared.cpp:76]: The include is needed for strcasecmp and _stricmp.
If you do a grep like `find . -name "*.cpp" | xargs grep "#include \<"`, you'll find many .cpp files include std header files.

File: src/hotspot/share/memory/metaspaceShared.cpp (hunk 2 above)
Comment (calvinccheung) [reply src/hotspot/share/memory/metaspaceShared.cpp:1821]: According to https://en.wikipedia.org/wiki/C_string_handling, strcasecmp is for POSIX, BSD, stricmp is for Windows.

File: test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/LambdaEagerInit.java (hunk 3 above)
Comment (calvinccheung) [reply test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/LambdaEagerInit.java:35]: Our current design is that every test under the `dynamicArchive` dir extends `DynamicArchiveTestBase` which depends on `sun.hotspot.WhiteBox`.

**Comment by calvinccheung** [comment]
Closing this pull request. I'll file another bug to address suggestions from @cl4es and @mlchung.

**Pull request closed without merging**